# Car Datasets
- Car Company Names: The manufacturer or brand of the car.
- Car Models: The specific name or series of the car.
- Engine Types: Information on engine specifications .
- CC/Battery Capacity: Engine displacement in cubic centimeters or battery capacity for electric cars.
- Horsepower (HP): The power output of the car's engine or motor.
- Top Speed: The maximum speed the car can achieve.
- 0-100 km/h Performance: The time it takes for the car to accelerate from 0 to 100 km/h.
- Price (in USD): The car's price listed in United States dollars.
- Fuel Type: Specifies whether the car uses petrol, diesel, electricity, or hybrid fuel systems.
- Seating Capacity: The number of passengers the car can accommodate.
- Torque: The rotational force the engine generates.

The objetive is to predict the price of the cars

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Reading data set

In [2]:
df = pd.read_csv("../Cars_Datasets_2025.csv", encoding="latin-1")

## EDA

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1218 entries, 0 to 1217
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Company Names              1218 non-null   str  
 1   Cars Names                 1218 non-null   str  
 2   Engines                    1218 non-null   str  
 3   CC/Battery Capacity        1215 non-null   str  
 4   HorsePower                 1218 non-null   str  
 5   Total Speed                1218 non-null   str  
 6   Performance(0 - 100 )KM/H  1212 non-null   str  
 7   Cars Prices                1218 non-null   str  
 8   Fuel Types                 1218 non-null   str  
 9   Seats                      1218 non-null   str  
 10  Torque                     1217 non-null   str  
dtypes: str(11)
memory usage: 104.8 KB


### Multiple null values found in the next columns
- CC/Battery Capacity: 3 nulls
- Performance(0 - 100 )KM/H: 6 nulls
- Torque: 1 nulls

In [4]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


## Company Names column

In [5]:
df["Company Names"].value_counts().sort_index()

Company Names
ASTON MARTIN          11
AUDI                  21
Acura                 27
BENTLEY                1
BMW                   41
Bugatti               10
Cadillac              20
Chevrolet             58
FERRARI                9
Ford                  54
GMC                   59
HONDA                  7
HYUNDAI               23
Jaguar Land Rover     40
Jeep                  19
KIA                   16
KIA                    1
Kia                   54
LAMBORGHINI           24
MAHINDRA               3
MARUTI SUZUKI          3
MERCEDES              21
Mazda                 87
Mitsubishi            60
NISSAN                10
Nissan               149
Peugeot               54
Porsche               96
ROLLS ROYCE           29
ROLLS ROYCE            1
TOYOTA                45
Tata Motors           29
Tesla                 10
Toyota                 5
VOLVO                  3
Volkswagen           109
Volvo                  9
Name: count, dtype: int64

### `Company Names`column. Multiple companys repeat with the same names
- Nissan / NISSAN
- Toyota / TOYOTA
- Kia / KIA / KIA
- ROLLS ROYCE / ROLLS ROYCE
- Volvo / volvo

## Car names column

In [6]:
def repeated_values(column):
    values = sorted(column.unique())
    values_set = set()
    j = 0
    for i in range(1, len(values)):
        if values[i].lower().strip() == values[j].lower().strip():
            if values[i] not in values_set:
                values_set.add(values[i])
        i = i + 1
        j = j + 1
    return values_set
            

In [7]:
values_set = repeated_values(df["Cars Names"])
print(values_set)

{'Panamera 4S '}


In [8]:
print(df["Cars Names"].value_counts().sort_index().to_string())

Cars Names
 GT 63 S                                   1
100NX                                      1
114D                                       1
114D SPORT LINE                            1
114i                                       1
116D                                       1
116D ADVANTAGE                             1
116D EFFICIENTDYNAMICS                     1
116i                                       1
116i EFFICIENTDYNAMICS                     1
118D                                       1
118D SPORT LINE                            1
118i                                       1
118i M SPORT                               1
118i SHADOW LINE                           1
118i SPORTLINE                             1
118i URBAN LINE                            1
118i XDRIVE                                1
1200                                       1
120D EFFICIENTDYNAMICS                     1
120D XDRIVE                                1
120i                                       1

### `Cars Names` 
The next value it's duplicated `Panamera 4S`. 

## Engines column

In [9]:
print(repeated_values(df["Engines"]))

{'ELECTRIC '}


In [10]:
print(df["Engines"].value_counts().sort_index().to_string())

Engines
0.36L Inline-2                                     1
0.66L Inline-3                                     1
0.7L 3-Cylinder                                    5
1.0L EcoBoost / 1.5L                               1
1.0L EcoBoost / 1.5L / 2.0L                        1
1.0L I3 Turbo                                      2
1.0L I3 Turbo / 1.5L I4 Turbo                      2
1.0L Inline-2 (Wankel)                             1
1.0L Turbo I3                                      1
1.0L Turbo I3 / 1.5L Turbo I4                      1
1.0L Turbo I3 / 1.6L Diesel                        1
1.0L Turbo Inline-3 / 2.0L                         1
1.0L Turbo Inline-4                                1
1.0L Turbocharged Inline-3                         4
1.0L,INLINE-3                                      1
1.1L / 1.5L MIVEC                                  1
1.2L Diesel I3                                     3
1.2L Hybrid                                        1
1.2L I4 Gas                           

### `Engines column` column
The next value it is duplicaded:
- ELECTRIC

In [11]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


## CC/Battery Capacity

In [12]:
print(df["CC/Battery Capacity"].value_counts().sort_index().to_string())

CC/Battery Capacity
-                                                    2
1,000 - 1,500 cc                                     1
1,000 - 2,000 cc                                     2
1,000 cc                                             2
1,100 - 1,500 cc                                     1
1,197 cc                                             3
1,199 cc                                            10
1,200 cc                                             9
1,298 cc                                             1
1,300 cc                                             4
1,308 cc                                             7
1,396 cc                                             1
1,462 cc                                             1
1,490 cc                                             4
1,496 cc                                             6
1,497 cc                                             3
1,498 cc                                             3
1,499 cc                                     

### The `CC/Battery Capacity` column has highly inconsistent formatting: 
Single values (`1,500 cc`), ranges (`1,000 - 1,500 cc`), combined combustion + electric powertrains (`1,598 cc (petrol) + Electric Motor`), mixed units within the same field (`1,600 cc / 13.8 kWh`) and null values expressed as `-`.
Before this can be used as a numeric feature, it will need to be standardized.

## HorsePower column

In [13]:
values_set = repeated_values(df["HorsePower"])
values_set

{'100 hp',
 '102 hp',
 '105 hp',
 '109 hp',
 '110 hp',
 '114 hp',
 '115 hp',
 '118 hp',
 '120 hp',
 '125 hp',
 '129 hp',
 '130 hp',
 '134 hp',
 '135 hp',
 '137 hp',
 '138 hp',
 '140 hp',
 '145 hp',
 '150 hp',
 '151 hp',
 '155 hp',
 '158 hp',
 '170 hp',
 '175 hp',
 '187 hp',
 '190 hp',
 '200 hp',
 '201 hp',
 '204 hp',
 '204 hp ',
 '205 hp',
 '210 hp',
 '247 hp',
 '250 hp',
 '255 hp',
 '261 hp',
 '268 hp',
 '270 hp',
 '276 hp',
 '280 hp',
 '284 hp',
 '300 hp',
 '325 hp',
 '332 hp',
 '335 hp',
 '340 hp',
 '350 hp',
 '375 hp',
 '379 hp',
 '400 hp',
 '420 hp',
 '434 hp',
 '440 hp',
 '453 hp',
 '455 hp',
 '473 hp',
 '48 hp',
 '500 hp',
 '550 hp',
 '600 hp',
 '670 hp',
 '690 hp',
 '85 hp'}

In [14]:
print(df["HorsePower"].value_counts().sort_index().to_string())

HorsePower
1,020 hp            2
1,288 hp            1
100 - 150 hp        1
100 HP              3
100 hp              7
1000 hp             1
102 - 150 hp        1
102 - 177 hp        1
102 - 204 hp        1
102 HP              1
102 hp              1
103 hp              1
104 HP              1
105 - 110 hp        1
105 - 210 hp        3
105 HP              2
105 hp              4
106 hp              1
108 hp              2
109 HP              1
109 hp              2
110 - 320 hp        1
110 HP              3
110 hp              6
113 hp              1
114 HP              1
114 hp              1
115 - 150 hp        2
115 - 300 hp        1
115 HP              3
115 hp              6
116 hp              4
1160 hp             1
117 - 148 hp        1
117 HP              1
118 HP              1
118 hp              1
118 hp / 110 hp     1
119 hp              1
120 - 240 hp        1
120 - 300 hp        1
120 HP              2
120 hp              9
121 hp              2
122 - 150 hp        1

### The `HorsePower` column is also highly inconsistent in format.
It mixes single values (`130 hp`, `310 hp`), ranges (`100 - 150 hp`,
`290-480 hp`, note the inconsistent spacing around the dash), inconsistent
casing (`hp` vs `HP`), dual values likely for different trims/engines
(`118 hp / 110 hp`), approximate values (`~170 hp`), qualified estimates
(`400 hp (est.)`), open-ended bounds (`Up to 1,000 hp`), and at least one
likely data-entry error (`2488 cc`, which is a displacement value mixed
into the horsepower column). Before this can be used as a numeric feature,
it will need standardization: strip units and symbols, normalize casing,
resolve ranges/dual values (e.g., using the midpoint or minimum), and
review outliers or misplaced values like the `cc` entry.

## Total spped column

In [15]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


In [16]:
print(df["Total Speed"].value_counts().sort_index().to_string())

Total Speed
100 km/h             2
105 km/h             1
110 km/h             4
120 km/h            15
125 km/h             1
130 km/h            15
138 km/h             1
140 km/h            16
145 km/h             9
148 km/h             1
150 km/h            10
155 km/h             5
160 km/h            44
162 km/h             1
165 km/h            12
168 km/h             1
170 km/h            47
175 km/h            24
177 km/h             7
180 km/h           129
183 km/h             1
185 km/h            18
186 km/h             2
187 km/h             1
190 km/h            92
191 km/h             2
193 km/h             1
195 km/h             8
196 km/h             1
200 km/h           139
201 km/h             1
204 km/h             1
205 km/h             9
209 km/h             5
210 km/h            85
211 km/h             1
213 km/h             5
215 km/h             8
216 km/h             2
217 km/h             4
220 km/h            48
221 km/h             1
222 km/h             1

### The `Total Speed` column needs to standardized.
It has to be remove `km/h` unit from all the entries. Also, one entry contains the indicator `(est.)`, which needs to be removed.

## Performance(0 - 100 )KM/H column

In [17]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


In [18]:
print(df["Performance(0 - 100 )KM/H"].value_counts().sort_index().to_string())

Performance(0 - 100 )KM/H
 6.1 sec               1
1.9 sec                1
10 sec                 2
10. 5 sec              1
10.0 sec              19
10.1 sec               1
10.2 sec              14
10.3 sec               5
10.4 sec               3
10.5 sec              38
10.6 sec               3
10.7 sec               4
10.8 sec               4
10.9 sec               4
11 sec                 3
11.0 sec              18
11.2 sec               6
11.5 sec              11
11.7 sec               1
11.9 sec               2
12 sec                 3
12.0 sec              17
12.2 sec               6
12.3 sec               1
12.4 sec               1
12.5 sec              13
12.8 sec               1
12.9 sec               2
13 sec                 3
13.0 sec               9
13.5 sec               6
13.7 sec               1
13.8 sec               1
14 sec                 1
14.0 sec               8
14.3 sec               1
14.5 sec               3
14.8 sec               2
15 sec                 3

### `Performance(0 - 100 )KM/H` column is inconsistent
In all entries needs to be removed any entra word like `sec` and `(GT Model)`. Also, it has to be removed any operations, ranges or special characteres `(9.0 - 12.0 sec, 7.2 sec / 5.4 sec, 8.0 ? 9.2 sec)`, so the column can be parse to float type

## Cars Prices column

In [19]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


In [20]:
print(df["Cars Prices"].value_counts().to_string())

Cars Prices
$35,000                    36
$30,000                    32
$50,000                    31
$55,000                    29
$25,000                    28
$45,000                    28
$40,000                    27
$28,000                    23
$60,000                    22
$70,000                    18
$20,000                    15
$18,000                    15
$85,000                    14
$32,000                    13
$22,000                    12
$20,000 - $25,000          12
$38,000                    11
$52,000                     9
$42,000                     9
$36,000                     9
$25,000 - $30,000           9
$30,000 - $35,000           9
$27,000                     8
$33,000                     8
$65,000                     8
$48,000                     7
$75,000                     7
$44,000                     7
$15,000                     7
$23,000                     7
$35,000 - $40,000           7
$18,000 - $22,000           7
$90,000                     

In [21]:
df[df["Cars Prices"] == "$3,200,000 "]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
32,ASTON MARTIN,VALKYRIE,V12,"6,496 cc",1160 hp,402 km/h,2.5 sec,"$3,200,000",Hybrid,2,900 Nm


### The `Car Prices` column is inconsistent
Needs to be removed characters like dollar sing `$`, comas `,`, unknown chracters `?`, any range `($5,000,000 - $7,000,000, 28,000 - $32,000)` and replace `(N/A (Concept Only)` with other value. As well, once cleand review outliers, there are cars that cost more than `$1,000,000`

## Fuel types column

In [22]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


In [23]:
print(df["Fuel Types"].value_counts().to_string())

Fuel Types
Petrol                     871
Diesel                     106
Electric                    97
Hybrid                      79
Petrol/Diesel               16
Petrol/Hybrid               16
Plug-in Hybrid               5
Diesel/Petrol                4
Hybrid (Petrol)              4
Hydrogen                     3
Petrol, Diesel               3
Petrol/AWD                   2
Petrol, Hybrid               2
plug in hyrbrid              1
Petrol/EV                    1
Hybrid/Electric              1
CNG/Petrol                   1
Hybrid/Petrol                1
Diesel Hybrid                1
Petrol (Hybrid)              1
Hybrid (Gas + Electric)      1
Gas / Hybrid                 1
Hybrid / Plug-in             1


### `Fuel types` column 
There are several values that are essentially the same but differ slightly, such as a comma versus a slash `(Petrol/Diesel, Petrol,Diesel)`. One entry mentions `AWD (All-Wheel Drive)`, which doesn't make sense in this column and should be removed. Each value in this column must be transformed into a unique numeric value.

## Seats column

In [24]:
df.head()

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
0,FERRARI,SF90 STRADALE,V8,3990 cc,963 hp,340 km/h,2.5 sec,"$1,100,000",plug in hyrbrid,2,800 Nm
1,ROLLS ROYCE,PHANTOM,V12,6749 cc,563 hp,250 km/h,5.3 sec,"$460,000",Petrol,5,900 Nm
2,Ford,KA+,1.2L Petrol,"1,200 cc",70-85 hp,165 km/h,10.5 sec,"$12,000-$15,000",Petrol,5,100 - 140 Nm
3,MERCEDES,GT 63 S,V8,"3,982 cc",630 hp,250 km/h,3.2 sec,"$161,000",Petrol,4,900 Nm
4,AUDI,AUDI R8 Gt,V10,"5,204 cc",602 hp,320 km/h,3.6 sec,"$253,290",Petrol,2,560 Nm


In [25]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1218 entries, 0 to 1217
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Company Names              1218 non-null   str  
 1   Cars Names                 1218 non-null   str  
 2   Engines                    1218 non-null   str  
 3   CC/Battery Capacity        1215 non-null   str  
 4   HorsePower                 1218 non-null   str  
 5   Total Speed                1218 non-null   str  
 6   Performance(0 - 100 )KM/H  1212 non-null   str  
 7   Cars Prices                1218 non-null   str  
 8   Fuel Types                 1218 non-null   str  
 9   Seats                      1218 non-null   str  
 10  Torque                     1217 non-null   str  
dtypes: str(11)
memory usage: 104.8 KB


In [26]:
print(df["Seats"].value_counts().to_string())

Seats
5       692
4       161
7       147
2       136
8        29
6        20
3        14
2+2       5
1         3
2–6       2
20        1
15        1
12        1
9         1
2–15      1
2–7       1
7–8       1
2–9       1
2–12      1


In [27]:
df[df["Seats"] == "12"]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
286,Nissan,NV1500,4.0L V6,3954 cc,261 hp,160 km/h,9.5 sec,"$33,000",Petrol,12,385 Nm


In [28]:
df[df["Seats"] == "15"]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
255,Nissan,Urvan,2.5L Turbo Diese,NaN,2488 cc,140 km/h,14.0 sec,"$28,000",Diesel,15,356 Nm


In [29]:
df[df["Seats"] == "20"]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
241,TOYOTA,COASTER,"4.0L,DIESEL","4,009 cc",134 hp,120 km/h,NaN,"$70,000",Diesel,20,500 Nm


In [30]:
df[df["Seats"] == "9"]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
327,Volkswagen,Transporter,2.0L I4 Turbo Diesel,1968 cc,102 - 204 hp,180 km/h,12.2 sec,"$35,000",Diesel,9,340 Nm


In [31]:
df[df["Seats"] == "2+2"]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
11,FERRARI,PORTOFINO,V8,"3,900 cc",592 hp,320 km/h,3.2 sec,"$210,000",Petrol,2+2,760 Nm
12,FERRARI,ROMA,V8,"3,900 cc",612 hp,320 km/h,3.4 sec,"$230,000",Petrol,2+2,760 Nm
15,FERRARI,PORTOFINO M,V8,"3,900 cc",612 hp,320 km/h,3.2 sec,"$220,000",Petrol,2+2,760 Nm
16,FERRARI,ROMA SPIDER,V8,"3,900 cc",612 hp,320 km/h,3.4 sec,"$240,000",Petrol,2+2,760 Nm
18,TOYOTA,TOYOTA 86,BOXER-4,"1,998 cc",205 hp,226 km/h,6.4 sec,"$27,000",Petrol,2+2,205 Nm


### `Seats column`
The unknown characters need to be removed `?` `(2?6)`. The cars with high quantity of seats seems to make sense. It's important to mention that `2+2` seat it's a valid scale for cars with two front seats and two small rear seats

In [32]:
print(df["Torque"].value_counts().to_string())

Torque
400 Nm               72
250 Nm               56
350 Nm               39
500 Nm               34
450 Nm               33
519 Nm               22
600 Nm               20
560 Nm               19
850 Nm               19
300 Nm               19
200 Nm               18
624 Nm               17
360 Nm               16
320 Nm               15
900 Nm               14
366 Nm               14
260 Nm               13
700 Nm               12
750 Nm               12
420 Nm               12
355 Nm               12
310 Nm               12
275 Nm               12
430 Nm               12
800 Nm               11
270 Nm               11
230 Nm               11
140 Nm               11
160 Nm               11
180 Nm               11
520 Nm               10
650 Nm               10
280 Nm               10
264 Nm               10
179 Nm               10
1600 Nm              10
380 Nm                9
240 Nm                9
870 Nm                9
550 Nm                9
170 Nm                9
265 Nm   

## `Torque` column
It needs to be removed the word `Nm`, commas, extra information `(GT Model)`, ranges `(200 - 400 Nm)`, special characters `(+, ?)`, and any range or operation `(100 - 180 Nm, 365 Nm / 480 Nm  )` also need to be resolved into a single numeric value.

## General duplicated rows

In [36]:
df[df.duplicated()]

,Company Names,Cars Names,Engines,CC/Battery Capacity,HorsePower,Total Speed,Performance(0 - 100 )KM/H,Cars Prices,Fuel Types,Seats,Torque
336,Volkswagen,Golf Cabriolet,1.2L I4 Turbo / 2.0L I4 Turbo,1197 cc / 1984 cc,105 - 210 hp,217 km/h,8.4 sec,"$30,000",Petrol,4,200 Nm
354,Volkswagen,Jetta Hybrid,1.4L I4 Turbo + Electric Motor,1395 cc,170 hp,210 km/h,8.6 sec,"$28,000",Hybrid,5,250 Nm
658,Tata Motors,Tiago EV,Permanent Magnet Synchronous,24 kWh,74 hp,120 km/h,5.7 sec,"$11,000",Electric,5,114 Nm
755,Chevrolet,Tahoe RST,5.3L V8 Gasoline,5300 cc,355 hp,180 km/h,7.4 sec,"$64,100",Petrol,7,519 Nm


# Conclusion
All 11 columns are stored as text (`str`) and require cleaning before any
numeric analysis or modeling is possible. Common issues across columns
include:

- Units embedded in values (`cc`, `hp`, `Nm`, `km/h`, `sec`)
- Currency symbols and thousands separators (`$`, `,`)
- Ranges and multi-value entries that need to be resolved into a single
  number (ex., `100 - 150 hp`, `365 Nm / 480 Nm`)
- Inconsistent casing and duplicate categories (`Nissan` vs `NISSAN`)
- Corrupted/unknown characters, likely from encoding issues (`?`)
- A few domain-specific edge cases: hybrid/electric combined values,
  `2+2` seating configuration, and outlier prices for rare or
  concept-only vehicles.

Also, there are 4 duplicated rows that have to be studied to be deleted
or keept

The next step is to clean each column into a proper numeric or
categorical type, so that descriptive statistics, distributions, and
correlations can be computed.